File: app/src/main/java/com/example/playback/SpotifyBroadCastReceiver.java
package com.example.playback;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.util.Log;

public class SpotifyBroadCastReceiver extends BroadcastReceiver {
    static final class BroadcastTypes {
        static final String SPOTIFY_PACKAGE = "com.spotify.music";
        static final String PLAYBACK_STATE_CHANGED = SPOTIFY_PACKAGE + ".playbackstatechanged";
        static final String QUEUE_CHANGED = SPOTIFY_PACKAGE + ".queuechanged";
        static final String METADATA_CHANGED = SPOTIFY_PACKAGE + ".metadatachanged";
    }
    private  static final String TAG = "SPOTIFY TESTING";

    @Override
    public void onReceive(Context context, Intent intent) {
        Log.v(TAG, "  on Receive");
        Log.v(TAG, "  DOES THIS WORK IDK???" );
        // This is sent with all broadcasts, regardless of type. The value is taken from
        // System.currentTimeMillis(), which you can compare to in order to determine how
        // old the event is.
        long timeSentInMs = intent.getLongExtra("timeSent", 0L);

        String action = intent.getAction();

        if (action.equals(BroadcastTypes.METADATA_CHANGED)) {
            String trackId = intent.getStringExtra("id");
            String artistName = intent.getStringExtra("artist");
            String albumName = intent.getStringExtra("album");
            String trackName = intent.getStringExtra("track");
            int trackLengthInSec = intent.getIntExtra("length", 0);

            Log.v(TAG, "  on Receive track changed to " + artistName);
            // Do something with extracted information...
        } else if (action.equals(BroadcastTypes.PLAYBACK_STATE_CHANGED)) {
            boolean playing = intent.getBooleanExtra("playing", false);
            int positionInMs = intent.getIntExtra("playbackPosition", 0);
            // Do something with extracted information
        } else if (action.equals(BroadcastTypes.QUEUE_CHANGED)) {
            // Sent only as a notification, your app may want to respond accordingly.
        } else if (action.equals(Intent.ACTION_BOOT_COMPLETED)){
            Log.v(TAG, "BOOT WORKED SO DOES IS RECEIVER!");
        }
    }
}

File: app/src/main/java/com/example/playback/SpotifyBroadCastReceiver.kt
package com.example.playback

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.util.Log


class SpotifyBroadCastReceiver : BroadcastReceiver() {
    internal object BroadcastTypes {
        const val SPOTIFY_PACKAGE = "com.spotify.music"
        const val PLAYBACK_STATE_CHANGED =  SPOTIFY_PACKAGE + "playbackstatechanged"
        const val QUEUE_CHANGED = SPOTIFY_PACKAGE + "queuechanged"
        const val METADATA_CHANGED = SPOTIFY_PACKAGE + "metadatachanged"
    }

    fun onReceive(context: Context?, intent: Intent) {
        Log.v(TAG, "  onReceive started")

        // This is sent with all broadcasts, regardless of type. The value is taken from
        // System.currentTimeMillis(), which you can compare to in order to determine how
        // old the event is.
        val timeSentInMs: Long = intent.getLongExtra("timeSent", 0L)

        //getting what kind of broad cast has been sent. 3 different types
        val action: String = intent.getAction()

        if (action == BroadcastTypes.METADATA_CHANGED) {
            val trackId: String = intent.getStringExtra("id")
            val artistName: String = intent.getStringExtra("artist")
            val albumName: String = intent.getStringExtra("album")
            val trackName: String = intent.getStringExtra("track")
            val trackLengthInSec: Int = intent.getIntExtra("length", 0)
            Log.v(TAG, "  on Receive track changed to" + artistName)
            // Do something with extracted information...
        } else if (action == BroadcastTypes.PLAYBACK_STATE_CHANGED) {
            val playing: Boolean = intent.getBooleanExtra("playing", false)
            val positionInMs: Int = intent.getIntExtra("playbackPosition", 0)
            // Do something with extracted information
        } else if (action == BroadcastTypes.QUEUE_CHANGED) { // Sent only as a notification, your app may want to respond accordingly.
        }
    }

    companion object {
        private const val TAG = "BroadCastReceiver"
    }
}
